include/pybind11/pybind11.h

```diff
@@ -885,11 +885,21 @@ class class_ : public detail::generic_type {
 
     template <typename... Extra>
     class_(handle scope, const char *name, const Extra &... extra) {
-        detail::type_record record;
+        using namespace detail;
+
+        // MI can only be specified via class_ template options, not constructor parameters
+        static_assert(
+            none_of<is_pyobject<Extra>...>::value || // no base class arguments, or:
+            (   constexpr_sum(is_pyobject<Extra>::value...) == 1 && // Exactly one base
+                constexpr_sum(is_base<options>::value...)   == 0 && // no template option bases
+                none_of<std::is_same<multiple_inheritance, Extra>...>::value), // no multiple_inheritance attr
```

Comment:
There's one annoying issue (MSVC related, as usual) with doing it that way: MSVC considers all the constexpr variables unused if only used in the `static_assert`, and so issues a warning which breaks the build, and so separate variables would need a `(void) (compiletime_bases, runtime_bases, needs_mi);` to silence it, which is a bit ugly.

I thought it might also end up being potentially less work for the compiler this way: typically the number of base classes is 0, so short-circuiting avoids needing to compute everything but the first one.